File: log-view-machine/java/src/main/java/com/logview/causality/Interpreter.java
package com.logview.causality;

import java.time.Instant;
import java.util.function.BiFunction;

/**
 * Interpreted state machine service (XState-congruent).
 * Uses CausalityTreeOps for forward/past/redo; send() is write, undo/redo delegate to tree.
 */
public final class Interpreter<T> {

    private final StateMachine<T> machine;
    private final CausalityTreeOps<T> tree;
    private boolean started;

    private Interpreter(StateMachine<T> machine, CausalityTreeOps<T> tree) {
        this.machine = machine;
        this.tree = tree;
    }

    public Interpreter<T> start() {
        if (!started) {
            started = true;
        }
        return this;
    }

    public Interpreter<T> stop() {
        started = false;
        return this;
    }

    /**
     * Send an event. Resolves transition from config; if found, writes new node (clears redo).
     * Event can be String (event type) or Map with "type" key.
     */
    public Interpreter<T> send(Object event) {
        if (!started) return this;
        CausalityNode<T> head = tree.getHead();
        String target = machine.getConfig().resolveTransition(head.getStateValue(), event);
        if (target == null) return this;
        CausalityNode<T> nextNode = new CausalityNode<>(
            target,
            head.getContext(),
            event,
            Instant.now(),
            head,
            null
        );
        tree.write(nextNode);
        return this;
    }

    /** Send with context update: assign (context, event) -> newContext. */
    public Interpreter<T> send(Object event, BiFunction<T, Object, T> assign) {
        if (!started) return this;
        CausalityNode<T> head = tree.getHead();
        String target = machine.getConfig().resolveTransition(head.getStateValue(), event);
        if (target == null) return this;
        T nextContext = assign.apply(head.getContext(), event);
        CausalityNode<T> nextNode = new CausalityNode<>(
            target,
            nextContext,
            event,
            Instant.now(),
            head,
            null
        );
        tree.write(nextNode);
        return this;
    }

    public StateSnapshot<T> getSnapshot() {
        CausalityNode<T> head = tree.getHead();
        return new StateSnapshot<>(
            head.getStateValue(),
            head.getContext(),
            head.getEvent()
        );
    }

    public String getValue() { return tree.getHead().getStateValue(); }
    public T getContext() { return tree.getHead().getContext(); }

    public boolean undo() { return tree.undo(); }
    public boolean redo() { return tree.redo(); }
    public boolean canUndo() { return tree.canUndo(); }
    public boolean canRedo() { return tree.canRedo(); }
    /** Undo until root (no events applied). Same as tri-leaf "backward". */
    public boolean undoToRoot() { return tree.undoToRoot(); }
    /** Alias for undoToRoot (tri-leaf backward action). */
    public boolean backward() { return tree.undoToRoot(); }
    public void pause() { tree.pause(); }
    public boolean resume() { return tree.resume(); }
    public boolean isPaused() { return tree.isPaused(); }

    /**
     * Manual state switch: write a node with event = OVERRIDE and the given state; context unchanged.
     */
    public Interpreter<T> setState(String targetState) {
        return setState(targetState, null);
    }

    /**
     * Manual state switch: write a node with event = OVERRIDE and the given state/context.
     * Goes through the same tree so undo-to-root includes overrides; auditable as override.
     */
    public Interpreter<T> setState(String targetState, T context) {
        if (!started) return this;
        CausalityNode<T> head = tree.getHead();
        T nextContext = context != null ? context : head.getContext();
        CausalityNode<T> nextNode = new CausalityNode<>(
            targetState,
            nextContext,
            MachineConfig.OVERRIDE_EVENT,
            Instant.now(),
            head,
            null
        );
        tree.write(nextNode);
        return this;
    }

    /** Create an interpreter with a plain causality tree. */
    public static <T> Interpreter<T> interpret(StateMachine<T> machine) {
        CausalityNode<T> initialNode = new CausalityNode<>(
            machine.getInitial(),
            machine.getContext(),
            null,
            Instant.now(),
            null,
            null
        );
        CausalityTree<T> tree = new CausalityTree<>(initialNode);
        return new Interpreter<>(machine, tree).start();
    }

    /**
     * Create an interpreter that uses the given tree (e.g. an ObservingCausalityTree for logging/metrics).
     * The tree must already have the initial node as head.
     */
    public static <T> Interpreter<T> interpret(StateMachine<T> machine, CausalityTreeOps<T> tree) {
        return new Interpreter<>(machine, tree).start();
    }
}

File: log-view-machine/core/src/main/kotlin/com/logview/causality/Interpreter.kt
package com.logview.causality

import java.time.Instant

/**
 * Interpreted state machine service (XState-congruent).
 * Uses CausalityTree for forward/past/redo; send() is write, undo/redo delegate to tree.
 */
class Interpreter<T> private constructor(
    private val machine: StateMachine<T>,
    private val tree: CausalityTree<T>
) {
    private var started = false

    fun start(): Interpreter<T> {
        if (started) return this
        started = true
        return this
    }

    fun stop(): Interpreter<T> {
        started = false
        return this
    }

    /**
     * Send an event. Resolves transition from config; if found, writes new node (clears redo).
     * Event can be String (event type) or Map with "type" key.
     */
    fun send(event: Any): Interpreter<T> {
        if (!started) return this
        val head = tree.getHead()
        val target = machine.config.resolveTransition(head.stateValue, event) ?: return this
        val nextNode = CausalityNode(
            stateValue = target,
            context = head.context,
            event = event,
            timestamp = Instant.now(),
            parent = head
        )
        tree.write(nextNode)
        return this
    }

    /** Send with context update: assign (context, event) -> newContext. */
    fun send(event: Any, assign: (T, Any) -> T): Interpreter<T> {
        if (!started) return this
        val head = tree.getHead()
        val target = machine.config.resolveTransition(head.stateValue, event) ?: return this
        val nextContext = assign(head.context, event)
        val nextNode = CausalityNode(
            stateValue = target,
            context = nextContext,
            event = event,
            timestamp = Instant.now(),
            parent = head
        )
        tree.write(nextNode)
        return this
    }

    fun getSnapshot(): StateSnapshot<T> {
        val head = tree.getHead()
        return StateSnapshot(
            value = head.stateValue,
            context = head.context,
            event = head.event,
            done = false
        )
    }

    @Suppress("UNUSED_PARAMETER")
    val value: String get() = tree.getHead().stateValue
    val context: T get() = tree.getHead().context

    fun undo(): Boolean = tree.undo()
    fun redo(): Boolean = tree.redo()
    fun canUndo(): Boolean = tree.canUndo()
    fun canRedo(): Boolean = tree.canRedo()
    /** Undo until root (no events applied). Same as tri-leaf "backward". */
    fun undoToRoot(): Boolean = tree.undoToRoot()
    /** Alias for undoToRoot (tri-leaf backward action). */
    fun backward(): Boolean = tree.undoToRoot()
    fun pause(): Unit = tree.pause()
    fun resume(): Boolean = tree.resume()
    fun isPaused(): Boolean = tree.isPaused()

    /**
     * Manual state switch: write a node with event = OVERRIDE and the given state/context.
     * Goes through the same tree so undo-to-root includes overrides; auditable as override.
     */
    fun setState(targetState: String, context: T? = null): Interpreter<T> {
        if (!started) return this
        val head = tree.getHead()
        val nextContext = context ?: head.context
        val nextNode = CausalityNode(
            stateValue = targetState,
            context = nextContext,
            event = OVERRIDE_EVENT,
            timestamp = Instant.now(),
            parent = head
        )
        tree.write(nextNode)
        return this
    }

    companion object {
        fun <T> interpret(machine: StateMachine<T>): Interpreter<T> {
            val initialNode = CausalityNode(
                stateValue = machine.initial,
                context = machine.context,
                event = null,
                timestamp = Instant.now(),
                parent = null
            )
            val tree = CausalityTree(initialNode)
            return Interpreter(machine, tree).start()
        }
    }
}

/** Convenience: interpret(machine) */
fun <T> interpret(machine: StateMachine<T>): Interpreter<T> =
    Interpreter.interpret(machine)
